lib_xua/src/core/mixer/mixer.xc

```diff
@@ -352,6 +335,63 @@ static inline void GetSamplesFromDevice(chanend c)
     }
 }
 
+#if (OUT_VOLUME_IN_MIXER && OUT_VOLUME_AFTER_MIX)
+#pragma unsafe arrays
+static inline void do_output_volume_control(int out_ch_index)
+{
+    int sample;
+    int mult;
+    int h;
+    unsigned l;
+
+    /* If mixer turned on sort out the channel mapping */
+    unsafe
+    {
+        /* Read index to sample from the map then Read the actual sample value */
+#if (MAX_MIX_COUNT > 0)
+        /* If mixer turned on sort out the channel mapping */
+        unsafe
+        {
+            /* Read index to sample from the map then Read the actual sample value */
+            sample = ptr_samples[samples_to_device_map[out_ch_index]];
+        }
+#else
+        unsafe
+        {
+            /* Read the actual sample value */
+            sample = ptr_samples[out_ch_index];
+        }
+#endif
+        mult = multOut[out_ch_index];
+    }
+
+    /* Do volume control processing */
+    {h, l} = macs(mult, sample, 0, 0);
+    h<<=3;              // Shift used to be done in audio thread but now done here incase of 32bit support
+#if (STREAM_FORMAT_OUTPUT_RESOLUTION_32BIT_USED == 1)
+    h |= (l >>29)& 0x7; // Note: This step is not required if we assume sample depth is 24bit (rather than 32bit)
+                        // Note: We need all 32bits for Native DSD
+#endif
+/* If mixer turned on sort out the channel mapping */
+#if (MAX_MIX_COUNT > 0)
+    /* If mixer turned on sort out the channel mapping */
+    unsafe
+    {
+        /* Read index to sample from the map then Read the actual sample value */
+        ptr_samples[samples_to_device_map[out_ch_index]] = h;
```

Comment:
does this value ever get used? Mixing is done, volume control is after the mix, so a new value will be stored here before this processed one is used?

Reply:
you're right! thanks for spotting this! I was doing it this way before introducing the new samples_to_device array to store the volume controlled samples. 
I guess the purpose of samples_to_device is not just to speed up the giving of samples to audio, but it is required to store a copy of the volume controlled output samples so we dont feed them back to the mixer. I'll add that to the comments